### PR TITLE
plat: intel: Add BL2 support for Stratix 10 SoC

### PR DESCRIPTION
**Pull request by thloh85-intel**
This adds BL2 support for Intel Stratix 10 SoC FPGA.
Functionality includes:
- Release and setup peripherals from reset
- Calibrate DDR
- ECC DDR Scrubbing
- Load FIP (bl31 and bl33)

Signed-off-by: Loh Tien Hock <tien.hock.loh@intel.com>

--
v2:
- Removed all deprecated dependencies
- Updated include in platform.mk to point to base only
- Removed stray files and codes
- Fixed ECC detection code

### CODE REVIEW

File: plat/intel/soc/stratix10/bl2_plat_setup.c [hunk 1]
```diff
@@ -0,0 +1,138 @@
...
+	boot_source = reverse_handoff_ptr.boot_source;
+
+	config_clkmgr_handoff(&reverse_handoff_ptr);
+	enable_nonsecure_access();
+	deassert_peripheral_reset();
+	config_hps_hs_before_warm_reset();
+
+	static console_16550_t console;
+	console_16550_register(PLAT_UART0_BASE, PLAT_UART_CLOCK, PLAT_BAUDRATE, &console);
+//	console_init(PLAT_UART0_BASE, PLAT_UART_CLOCK, PLAT_BAUDRATE);
```
Comment (antonio-nino-diaz-arm) [inline plat/intel/soc/stratix10/bl2_plat_setup.c]: Remove this commented-out line. That function is deprecated, anyway, ``console_16550_register()`` is enough.

File: plat/intel/soc/stratix10/bl2_plat_setup.c [hunk 2]
```diff
@@ -0,0 +1,138 @@
...
+#include <arch_helpers.h>
+#include <drivers/arm/gicv2.h>
+
+#include <drivers/generic_delay_timer.h>
+#include <drivers/console.h>
+#include <drivers/ti/uart/uart_16550.h>
+#include <common/bl_common.h>
+#include <common/debug.h>
+#include <common/desc_image_load.h>
+#include <lib/libc/errno.h>
```
Comment (antonio-nino-diaz-arm) [inline plat/intel/soc/stratix10/bl2_plat_setup.c]: ``errno.h`` is a libc header, don't use the full path for this one.

File: plat/intel/soc/stratix10/aarch64/plat_helpers.S [hunk 3]
```diff
@@ -0,0 +1,128 @@
...
+	 * Clobber list : x1, x2
+	 * ---------------------------------------------
+	 */
+func plat_crash_console_putc
+	mov_imm x1, PLAT_UART0_BASE
+	b	console_16550_core_putc
+endfunc plat_crash_console_putc
+
+func plat_crash_console_flush
+	ret
```
Comment (antonio-nino-diaz-arm) [inline plat/intel/soc/stratix10/aarch64/plat_helpers.S]: This should return 0.

File: plat/intel/soc/stratix10/soc/s10_memory_controller.c [hunk 4]
```diff
@@ -0,0 +1,402 @@
...
+#define DDR_READ_LATENCY_DELAY 40
+#define MAX_MEM_CAL_RETRY		3
+#define PRE_CALIBRATION_DELAY		1
+#define POST_CALIBRATION_DELAY		1
+#define TIMEOUT_EMIF_CALIBRATION	100
+#define CLEAR_EMIF_DELAY		50000
+#define CLEAR_EMIF_TIMEOUT		0x100000
+#define TIMEOUT_INT_RESP		10000
+
+#define DDR_CONFIG(A, B, C, R)	((A<<24)|(B<<16)|(C<<8)|R)
```
Comment (antonio-nino-diaz-arm) [inline plat/intel/soc/stratix10/soc/s10_memory_controller.c]: You should add parentheses to the arguments.

**Comment by antonio-nino-diaz-arm** [comment]
Also, when you are done with the changes, squash all the commits into one.

File: plat/intel/soc/stratix10/aarch64/platform_common.c [hunk 5]
```diff
@@ -0,0 +1,109 @@
...
+ * SPDX-License-Identifier: BSD-3-Clause
+ */
+
+#include <arch_helpers.h>
+#include <lib/xlat_tables/xlat_tables.h>
+#include <lib/mmio.h>
+#include <platform_def.h>
+
+/* Table of regions to map using the MMU.  */
+static const mmap_region_t plat_mmap[] = {
```
[the remark this answers is not in the document]
Comment (thloh85-intel) [reply plat/intel/soc/stratix10/aarch64/platform_common.c]: This is no longer useful after moving to setup_page_tables and enable_mmu_elx(). I'll be removing this.

**Comment by thloh85-intel** [comment]
I've updated the requested change and squashed a few bugs. Please help review the latest commit. Thanks in advance 👍

**Pull request merged**